File: fastapi/routes/control.py
import os
from fastapi import APIRouter, Header, HTTPException

router = APIRouter(prefix="/control")

TRADING_ENABLED = {"value": True}
API_KEY = os.getenv("API_KEY")

def check_key(x_api_key: str | None):
    if not API_KEY:
        raise HTTPException(status_code=500, detail="API_KEY not set")
    if x_api_key != API_KEY:
        raise HTTPException(status_code=401, detail="Unauthorized")

@router.post("")
def control(action: dict, x_api_key: str | None = Header(default=None)):
    check_key(x_api_key)
    act = (action or {}).get("action")
    if act == "start":
        TRADING_ENABLED["value"] = True
    elif act == "stop":
        TRADING_ENABLED["value"] = False
    elif act != "status":
        raise HTTPException(status_code=400, detail="action must be start|stop|status")
    return {"ok": True, "trading_enabled": TRADING_ENABLED["value"]}

File: stage9-node/routes/control.js
import express from "express";
const router = express.Router();

function requireKey(req, res, next) {
  const k = req.header("X-API-Key");
  if (!process.env.API_KEY) return res.status(500).json({ ok:false, error:"API_KEY not set" });
  if (k !== process.env.API_KEY) return res.status(401).json({ ok:false, error:"Unauthorized" });
  next();
}

// in-memory flag
globalThis.__TRADING_ENABLED__ = (globalThis.__TRADING_ENABLED__ === undefined) ? true : globalThis.__TRADING_ENABLED__;

router.post("/", requireKey, (req, res) => {
  const { action } = req.body || {};
  if (action === "start") globalThis.__TRADING_ENABLED__ = true;
  else if (action === "stop") globalThis.__TRADING_ENABLED__ = false;
  else if (action !== "status") return res.status(400).json({ ok:false, error:"action must be start|stop|status" });
  res.json({ ok:true, trading_enabled: globalThis.__TRADING_ENABLED__ });
});

export default router;
